src/peergos/shared/user/BtreeImpl.java

```diff
@@ -66,21 +65,27 @@ public BtreeImpl(MutablePointers mutable, ContentAddressedStorage dht) {
     }
 
     @Override
-    public CompletableFuture<Boolean> put(SigningPrivateKeyAndPublicHash writer, byte[] mapKey, Multihash value) {
+    public CompletableFuture<Boolean> put(SigningPrivateKeyAndPublicHash writer, byte[] mapKey, MaybeMultihash existing, Multihash value) {
         PublicKeyHash publicWriterKey = writer.publicKeyHash;
         CompletableFuture<CommittedWriterData> lock = new CompletableFuture<>();
 
         return addToQueue(publicWriterKey, lock)
                 .thenCompose(committed -> {
                     WriterData holder = committed.props;
-                    MaybeMultihash btreeRootHash = holder.btree.isPresent() ? MaybeMultihash.of(holder.btree.get()) : MaybeMultihash.EMPTY();
+                    MaybeMultihash btreeRootHash = holder.btree.isPresent() ? MaybeMultihash.of(holder.btree.get()) : MaybeMultihash.empty();
                     return MerkleBTree.create(publicWriterKey, btreeRootHash, dht)
-                            .thenCompose(btree -> btree.put(publicWriterKey, mapKey, value))
+                            .thenCompose(btree -> btree.put(publicWriterKey, mapKey, existing, value))
                             .thenApply(newRoot -> LOGGING ? log(newRoot, "BTREE.put (" + ArrayOps.bytesToHex(mapKey)
                                     + ", " + value + ") => CAS(" + btreeRootHash + ", " + newRoot + ")") : newRoot)
                             .thenCompose(newBtreeRoot -> holder.withBtree(newBtreeRoot)
                                     .commit(writer, committed.hash, mutable, dht, lock::complete))
-                            .thenApply(x -> true);
+                            .thenApply(x -> true)
+                            .exceptionally(e -> {
+                                lock.complete(committed);
+                                if (e instanceof RuntimeException)
```

Comment:
not sure about this. runtimeexceptions in gwt not supported.

Reply:
gwt compiles it fine, haven't tested in the web ui though.